File: sdks/typescript/pmxt/feed-client.ts
/**
 * Feed client — CCXT-compatible method names over /api/feeds/* endpoints.
 *
 * Usage:
 *   const feed = new FeedClient('chainlink', { pmxtApiKey: '...' });
 *   const ticker = await feed.fetchTicker('BTC/USD');
 */

import { resolvePmxtBaseUrl } from "./constants.js";
import { PmxtError } from "./errors.js";

export interface Ticker {
    symbol: string;
    info: any;
    timestamp: number | undefined;
    datetime: string | undefined;
    high: number | undefined;
    low: number | undefined;
    bid: number | undefined;
    bidVolume: number | undefined;
    ask: number | undefined;
    askVolume: number | undefined;
    vwap: number | undefined;
    open: number | undefined;
    close: number | undefined;
    last: number | undefined;
    previousClose: number | undefined;
    change: number | undefined;
    percentage: number | undefined;
    average: number | undefined;
    quoteVolume: number | undefined;
    baseVolume: number | undefined;
    indexPrice: number | undefined;
    markPrice: number | undefined;
}

export type Tickers = Record<string, Ticker>;
export type OHLCV = [number, number, number, number, number, number];

export interface Market {
    id: string;
    symbol: string;
    base: string;
    quote: string;
    active: boolean;
    type: string;
    info: any;
}

export interface OracleRound {
    feed: string;
    roundId: string;
    answer: number;
    startedAt: number;
    updatedAt: number;
    answeredInRound: string;
    decimals: number;
    description?: string;
}

export interface FeedClientOptions {
    pmxtApiKey?: string;
    baseUrl?: string;
}

export class FeedClient {
    private readonly feedName: string;
    private readonly baseUrl: string;
    private readonly headers: Record<string, string>;

    constructor(feedName: string, options: FeedClientOptions = {}) {
        this.feedName = feedName;
        const resolved = resolvePmxtBaseUrl({
            baseUrl: options.baseUrl,
            pmxtApiKey: options.pmxtApiKey,
        });
        this.baseUrl = resolved.baseUrl;
        this.headers = {
            ...(resolved.pmxtApiKey ? { 'Authorization': `Bearer ${resolved.pmxtApiKey}` } : {}),
        };
    }

    async loadMarkets(): Promise<Record<string, Market>> {
        return this.get<Record<string, Market>>('loadMarkets', {});
    }

    async fetchTicker(symbol: string): Promise<Ticker> {
        return this.get<Ticker>('fetchTicker', { symbol });
    }

    async fetchTickers(symbols?: string[]): Promise<Tickers> {
        const params: Record<string, unknown> = {};
        if (symbols) params.symbols = symbols.join(',');
        return this.get<Tickers>('fetchTickers', params);
    }

    async fetchOHLCV(symbol: string, timeframe: string = '1h', since?: number, limit?: number): Promise<OHLCV[]> {
        return this.get<OHLCV[]>('fetchOHLCV', { symbol, timeframe, since, limit });
    }

    async fetchOracleRound(feed: string): Promise<OracleRound> {
        return this.get<OracleRound>('fetchOracleRound', { feed });
    }

    async fetchOracleHistory(feed: string, limit?: number): Promise<OracleRound[]> {
        const params: Record<string, unknown> = { feed };
        if (limit !== undefined) params.limit = limit;
        return this.get<OracleRound[]>('fetchOracleHistory', params);
    }

    async fetchHistoricalPrices(symbol: string, opts?: {
        fromTimestamp?: number;
        untilTimestamp?: number;
        maxSize?: number;
        order?: 'asc' | 'desc';
    }): Promise<Ticker[]> {
        return this.get<Ticker[]>('fetchHistoricalPrices', { symbol, ...opts });
    }

    private async get<T>(method: string, params: Record<string, unknown>): Promise<T> {
        const qs = Object.entries(params)
            .filter(([, v]) => v !== undefined && v !== null)
            .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(String(v))}`)
            .join('&');

        const url = `${this.baseUrl}/api/feeds/${this.feedName}/${method}${qs ? '?' + qs : ''}`;

        const response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            const body = await response.json().catch(() => ({}));
            throw new PmxtError(body.error || response.statusText);
        }

        const json = await response.json() as { success: boolean; data: T; error?: string };
        if (!json.success) {
            throw new PmxtError(json.error || 'Unknown feed error');
        }
        return json.data;
    }
}

File: sdks/python/pmxt/feed_client.py
"""
Feed client — CCXT-compatible method names over /api/feeds/* endpoints.

Usage:
    from pmxt.feed_client import FeedClient
    feed = FeedClient('chainlink', pmxt_api_key='...')
    ticker = feed.fetch_ticker('BTC/USD')
"""

import os
import urllib.request
import urllib.parse
import urllib.error
import json
from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any, Tuple


@dataclass(frozen=True)
class Ticker:
    symbol: str
    info: Any
    timestamp: Optional[int] = None
    datetime: Optional[str] = None
    high: Optional[float] = None
    low: Optional[float] = None
    bid: Optional[float] = None
    bid_volume: Optional[float] = None
    ask: Optional[float] = None
    ask_volume: Optional[float] = None
    vwap: Optional[float] = None
    open: Optional[float] = None
    close: Optional[float] = None
    last: Optional[float] = None
    previous_close: Optional[float] = None
    change: Optional[float] = None
    percentage: Optional[float] = None
    average: Optional[float] = None
    quote_volume: Optional[float] = None
    base_volume: Optional[float] = None
    index_price: Optional[float] = None
    mark_price: Optional[float] = None


OHLCV = Tuple[float, float, float, float, float, float]


@dataclass(frozen=True)
class Market:
    id: str
    symbol: str
    base: str
    quote: str
    active: bool
    type: str
    info: Any = field(default_factory=dict)


@dataclass(frozen=True)
class OracleRound:
    feed: str
    round_id: str
    answer: float
    started_at: int
    updated_at: int
    answered_in_round: str
    decimals: int
    description: Optional[str] = None


HOSTED_URL = "https://api.pmxt.dev"
LOCAL_URL = "http://localhost:3847"


def _resolve_base_url(pmxt_api_key: str) -> str:
    env_url = os.environ.get("PMXT_BASE_URL")
    if env_url:
        return env_url
    if pmxt_api_key:
        return HOSTED_URL
    return LOCAL_URL


class FeedClient:
    def __init__(
        self,
        feed_name: str,
        *,
        pmxt_api_key: Optional[str] = None,
        base_url: Optional[str] = None,
    ) -> None:
        self._feed_name = feed_name
        api_key = pmxt_api_key or os.environ.get("PMXT_API_KEY", "")
        self._base_url = base_url or _resolve_base_url(api_key)
        self._headers: Dict[str, str] = {}
        if api_key:
            self._headers["Authorization"] = f"Bearer {api_key}"

    def load_markets(self) -> Dict[str, Market]:
        data = self._get("loadMarkets", {})
        return {
            k: Market(
                id=v["id"],
                symbol=v["symbol"],
                base=v["base"],
                quote=v["quote"],
                active=v.get("active", True),
                type=v.get("type", "spot"),
                info=v.get("info", {}),
            )
            for k, v in data.items()
        }

    def fetch_ticker(self, symbol: str) -> Ticker:
        data = self._get("fetchTicker", {"symbol": symbol})
        return self._to_ticker(data)

    def fetch_tickers(self, symbols: Optional[List[str]] = None) -> Dict[str, Ticker]:
        params: Dict[str, Any] = {}
        if symbols:
            params["symbols"] = ",".join(symbols)
        data = self._get("fetchTickers", params)
        return {k: self._to_ticker(v) for k, v in data.items()}

    def fetch_ohlcv(
        self,
        symbol: str,
        timeframe: str = "1h",
        since: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> List[OHLCV]:
        params: Dict[str, Any] = {"symbol": symbol, "timeframe": timeframe}
        if since is not None:
            params["since"] = since
        if limit is not None:
            params["limit"] = limit
        data = self._get("fetchOHLCV", params)
        return [tuple(row) for row in data]

    def fetch_oracle_round(self, feed: str) -> OracleRound:
        data = self._get("fetchOracleRound", {"feed": feed})
        return self._to_oracle_round(data)

    def fetch_oracle_history(
        self, feed: str, limit: Optional[int] = None
    ) -> List[OracleRound]:
        params: Dict[str, Any] = {"feed": feed}
        if limit is not None:
            params["limit"] = limit
        data = self._get("fetchOracleHistory", params)
        return [self._to_oracle_round(r) for r in data]

    def fetch_historical_prices(
        self,
        symbol: str,
        *,
        from_timestamp: Optional[int] = None,
        until_timestamp: Optional[int] = None,
        max_size: Optional[int] = None,
        order: Optional[str] = None,
    ) -> List[Ticker]:
        params: Dict[str, Any] = {"symbol": symbol}
        if from_timestamp is not None:
            params["fromTimestamp"] = from_timestamp
        if until_timestamp is not None:
            params["untilTimestamp"] = until_timestamp
        if max_size is not None:
            params["maxSize"] = max_size
        if order is not None:
            params["order"] = order
        data = self._get("fetchHistoricalPrices", params)
        return [self._to_ticker(r) for r in data]

    def _get(self, method: str, params: Dict[str, Any]) -> Any:
        filtered = {k: v for k, v in params.items() if v is not None}
        qs = urllib.parse.urlencode(filtered) if filtered else ""
        url = f"{self._base_url}/api/feeds/{self._feed_name}/{method}"
        if qs:
            url += f"?{qs}"

        req = urllib.request.Request(url, headers=self._headers)
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                body = json.loads(resp.read())
        except urllib.error.HTTPError as e:
            body = json.loads(e.read()) if e.fp else {}
            msg = body.get("error", e.reason)
            raise RuntimeError(f"Feed API error ({e.code}): {msg}") from e

        if not body.get("success"):
            raise RuntimeError(f"Feed API error: {body.get('error', 'unknown')}")
        return body["data"]

    @staticmethod
    def _to_ticker(raw: Dict[str, Any]) -> Ticker:
        return Ticker(
            symbol=raw["symbol"],
            info=raw.get("info"),
            timestamp=raw.get("timestamp"),
            datetime=raw.get("datetime"),
            high=raw.get("high"),
            low=raw.get("low"),
            bid=raw.get("bid"),
            bid_volume=raw.get("bidVolume"),
            ask=raw.get("ask"),
            ask_volume=raw.get("askVolume"),
            vwap=raw.get("vwap"),
            open=raw.get("open"),
            close=raw.get("close"),
            last=raw.get("last"),
            previous_close=raw.get("previousClose"),
            change=raw.get("change"),
            percentage=raw.get("percentage"),
            average=raw.get("average"),
            quote_volume=raw.get("quoteVolume"),
            base_volume=raw.get("baseVolume"),
            index_price=raw.get("indexPrice"),
            mark_price=raw.get("markPrice"),
        )

    @staticmethod
    def _to_oracle_round(raw: Dict[str, Any]) -> OracleRound:
        return OracleRound(
            feed=raw["feed"],
            round_id=raw["roundId"],
            answer=raw["answer"],
            started_at=raw["startedAt"],
            updated_at=raw["updatedAt"],
            answered_in_round=raw["answeredInRound"],
            decimals=raw["decimals"],
            description=raw.get("description"),
        )
